Implement the reconciliation loop for GCPManagedMachinePool (GKE Part 4)

richardchen331 opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This change adds the reconciliation logic for the GCPManagedControlPlane.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Relates https://github.com/kubernetes-sigs/cluster-api-provider-gcp/issues/512

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add the reconciliation logic for **GCPManagedMachinePool**
```

cpanato commented:
/ok-to-test

richardcase commented on config/manager/manager.yaml:
```diff
@@ -21,6 +21,7 @@ spec:
       containers:
       - args:
         - --leader-elect
+        - --feature-gates=GKE=${EXP_CAPG_GKE:=false}
```
Can we remove this until PR5?

richardchen331 replied:
removed

richardcase commented on exp/controllers/gcpmanagedmachinepool_controller.go:
```diff
@@ -29,33 +57,320 @@ import (
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.2/pkg/reconcile
-func (r *GCPManagedMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+func (r *GCPManagedMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
```
We'll need permissions for MachinePool, Cluster and GCPManagedCluster, GCPManagedControlPlane

richardchen331 replied:
updated

richardcase commented on feature/feature.go:
```diff
@@ -20,12 +20,16 @@ import (
-// // owner: @username
-// // alpha: v1.X
-// MyFeature featuregate.Feature = "MyFeature".
+	// Every capg-specific feature gate should add method here following this template:
```
We can remove this as it will be added in PR5

richardcase replied:
I'm also wondering do we need to worry about the CAPI machinepool feature flag?

richardcase replied:
But we can worry about that in the next PR :)

richardchen331 replied:
change reverted :)

richardcase commented on main.go:
```diff
@@ -143,56 +147,117 @@ func main() {
-	if err = (&infrav1beta1.GCPCluster{}).SetupWebhookWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create webhook", "webhook", "GCPCluster")
-		os.Exit(1)
+	if feature.Gates.Enabled(feature.GKE) {
```
We'll do this in the next PR when we bring it all together.

richardchen331 replied:
reverted

richardcase commented on main.go:
```diff
@@ -143,56 +147,117 @@ func main() {
+	}
+
+	if feature.Gates.Enabled(feature.GKE) {
+		if err := (&infrav1exp.GCPManagedCluster{}).SetupWebhookWithManager(mgr); err != nil {
```
Same comment as above with enabling the GKE controllers in the next PR

richardchen331 replied:
reverted

richardcase reviewed (commented):
Thanks @richardchen331 🎉

richardchen331 commented on cloud/scope/managedmachinepool.go:
```diff
@@ -0,0 +1,222 @@
+	return parts[6], parts[8], parts[10]
+}
+
+// ConvertToSdkNodePool convertsSetReplicas a node pool to format that is used by GCP SDK.
```
fixed

richardchen331 commented on cloud/scope/managedmachinepool.go:
```diff
@@ -0,0 +1,222 @@
+		},
+	}
+	if nodePool.Spec.NodeVersion != nil {
+		sdkNodePool.Version = *nodePool.Spec.NodeVersion
```
make sense, updated

richardchen331 commented on cloud/scope/managedmachinepool.go:
```diff
@@ -0,0 +1,222 @@
+// ParseInstanceGroupURL parses an instance group URL.
+func ParseInstanceGroupURL(url string) (project string, zone string, mig string) {
+	parts := strings.Split(url, "/")
+	if len(parts) < 11 {
```
updated

richardchen331 commented on cloud/scope/managedcontrolplane.go:
```diff
@@ -171,18 +173,33 @@ func (s *ManagedControlPlaneScope) GetCredential() *Credential {
+	return s.AllNodePools, nil
+}
+
+func parseLocation(location string) string {
```
updated

richardcase reviewed (commented):
Fantastic @richardchen331

richardchen331 commented on cloud/scope/managedcontrolplane.go:
```diff
@@ -171,19 +176,36 @@ func (s *ManagedControlPlaneScope) GetCredential() *Credential {
+// GetAllNodePools gets all node pools for the control plane.
+func (s *ManagedControlPlaneScope) GetAllNodePools(ctx context.Context) ([]infrav1exp.GCPManagedMachinePool, []clusterv1exp.MachinePool, error) {
+	if s.AllManagedMachinePools == nil || len(s.AllManagedMachinePools) == 0 {
+		opt1 := client.InNamespace(s.GCPManagedControlPlane.Namespace)
```
updated

richardchen331 commented on cloud/services/container/clusters/kubeconfig.go:
```diff
@@ -198,7 +198,7 @@ func (s *Service) updateCAPIKubeconfigSecret(ctx context.Context, configSecret *
 
 func (s *Service) getKubeConfigContextName(isUser bool) string {
-	contextName := fmt.Sprintf("gke_%s_%s_%s", s.scope.GCPManagedControlPlane.Spec.Project, s.scope.GCPManagedControlPlane.Spec.Location, s.scope.GCPManagedControlPlane.Name)
+	contextName := fmt.Sprintf("gke_%s_%s_%s", s.scope.GCPManagedControlPlane.Spec.Project, s.scope.GCPManagedControlPlane.Spec.Location, s.scope.GCPManagedControlPlane.Spec.ClusterName)
```
sg! updated

richardchen331 commented on cloud/services/container/clusters/reconcile.go:
```diff
@@ -190,18 +215,14 @@ func (s *Service) describeCluster(ctx context.Context) (*containerpb.Cluster, er
+	nodePools, machinePools, _ := s.scope.GetAllNodePools(ctx)
 	cluster := &containerpb.Cluster{
-		Name:    s.scope.GCPManagedControlPlane.Name,
+		Name:    s.scope.GCPManagedControlPlane.Spec.ClusterName,
```
agreed, updated

richardchen331 commented on cloud/services/container/nodepools/reconcile.go:
```diff
@@ -0,0 +1,330 @@
+		return ctrl.Result{}, err
+	}
+	if nodePool == nil {
+		log.Info("Node pool not found, creating")
```
done

richardchen331 commented on cloud/services/container/nodepools/reconcile.go:
```diff
@@ -0,0 +1,330 @@
+		log.Info("Node pool version/image updating in progress")
+		s.scope.GCPManagedMachinePool.Status.Ready = true
+		conditions.MarkTrue(s.scope.ConditionSetter(), infrav1exp.GKEMachinePoolUpdatingCondition)
+		return ctrl.Result{}, nil
```
you're right, we should requeue here

richardchen331 commented on cloud/services/container/nodepools/reconcile.go:
```diff
@@ -0,0 +1,330 @@
+		log.Info("Node pool size updating in progress")
+		s.scope.GCPManagedMachinePool.Status.Ready = true
+		conditions.MarkTrue(s.scope.ConditionSetter(), infrav1exp.GKEMachinePoolUpdatingCondition)
+		return ctrl.Result{}, nil
```
make sense, updated

richardchen331 commented on cloud/services/container/nodepools/reconcile.go:
```diff
@@ -0,0 +1,330 @@
+	}
+
+	switch nodePool.Status {
+	case containerpb.NodePool_PROVISIONING:
```
i think GKE doesn't allow you to perform any operation (e.g. delete) while one is still in progress (e.g. provisioning), but we should requeue here.

richardchen331 commented on cloud/services/container/nodepools/reconcile.go:
```diff
@@ -0,0 +1,330 @@
+	switch nodePool.Status {
+	case containerpb.NodePool_PROVISIONING:
+		log.Info("Node pool provisioning in progress")
+		return ctrl.Result{}, nil
```
updated

richardchen331 commented on exp/api/v1beta1/gcpmanagedmachinepool_types.go:
```diff
@@ -30,13 +30,13 @@ const (
+	// NodeCount represents the initial number of nodes for the pool.
 	// In regional or multi-zonal clusters, this is the number of nodes per zone.
-	InitialNodeCount int32 `json:"initialNodeCount"`
+	NodeCount int32 `json:"nodeCount"`
```
my plan is to get an MVP working and adding more features after the initial set of PRs are merged. let me know if that make sense :)

richardcase replied:
I would agree with this approach for most features. But I wonder as this is a machinepool should we have scaling from the beginning? Feels like we should but happy to go with what you feel is best,

richardcase commented:
@richardchen331 - sorry for the late replay. I left 1 reply around auto-scaling. Apart from that, I think this looks good. Could you rebase and squash commits?

richardchen331 commented:
> @richardchen331 - sorry for the late replay. I left 1 reply around auto-scaling. Apart from that, I think this looks good. Could you rebase and squash commits?

Thanks for the review! I rebased and squashed the commits (only left one additional for the autoscaling change), and add auto scaling support. Please take another look when you have time. I can squash again after it's approved :)

richardchen331 commented:
/test pull-cluster-api-provider-gcp-e2e-test

richardcase commented:
@richardchen331 - this looks great, and thanks for adding the autoscaling options initially. I think this will help. From my side, i have no further comments on the code.

Can you squash and handle the conflicting `go.mod`? 

I will then approve :)

richardchen331 commented:
> @richardchen331 - this looks great, and thanks for adding the autoscaling options initially. I think this will help. From my side, i have no further comments on the code.
> 
> Can you squash and handle the conflicting `go.mod`?
> 
> I will then approve :)

Sure. I rebased and squashed the commits. Could you help approve the PR? Thanks Richard!

richardcase commented:
/override pull-cluster-api-provider-gcp-apidiff

cpanato reviewed (approved):
nice work

/approve
